src/app/icd/client/ICDClientStorage.h

```diff
@@ -0,0 +1,95 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "ICDClientInfo.h"
+#include <crypto/CHIPCryptoPAL.h>
+#include <lib/core/CHIPConfig.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/ScopedNodeId.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/CommonIterator.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * The ICDClientStorage class is an abstract interface that defines the operations
+ * for storing, retrieving and deleting ICD client information in persistent storage.
+ */
+class ICDClientStorage
+{
+public:
+    using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;
+
+    virtual ~ICDClientStorage() = default;
+
+    /**
+     * Iterate through persisted ICD Client Info
+     *
+     * @return A valid iterator on success. Use CommonIterator accessor to retrieve ICDClientInfo
+     */
+    virtual ICDClientInfoIterator * IterateICDClientInfo() = 0;
+
+    /**
+     * Called during ICD device registration in commissioning, commissioner/controller
+     * provides raw key data, the shared key handle in clientInfo is updated based upon raw key data
+     *
+     * @param[inout] aICDClientInfo the ICD Client information to be updated with keyData and be saved
+     * @param[in] aKeyData raw key data provided by application
+     */
+    virtual CHIP_ERROR SetKey(ICDClientInfo & clientInfo, const ByteSpan keyData) = 0;
+
+    /**
+     * Store updated ICD ClientInfo to storage when ICD registration completes or check-in message
+     * comes.
+     *
+     * @param[in] aICDClientInfo the updated ICD Client Info.
+     */
+    virtual CHIP_ERROR StoreEntry(const ICDClientInfo & clientInfo) = 0;
+
+    /**
+     * Delete ICD Client persistent information associated with the specified scoped node Id.
+     * when ICD device is unpaired/removed, the corresponding entry in ICD storage is removed.
+     * @param aPeerNodeId scoped node with peer node id and fabric index
+     */
+    virtual CHIP_ERROR DeleteEntry(const ScopedNodeId & peerNodeId) = 0;
+
+    /**
+     * Remove all ICDClient persistent informations associated with the specified
+     * fabric index.  If no entries for the fabric index exist, this is a no-op
+     * and is considered successful.
+     * When the whole fabric is removed, all entries from persistent storage in current fabric index is removed.
+     *
+     * @param[in] fabricIndex the index of the fabric for which to remove ICDClient persistent information
+     */
+    virtual CHIP_ERROR DeleteAllEntries(FabricIndex fabricIndex) = 0;
```

Comment:
Fine for now, but I am having a hard time imagining how this will work in practice; on controllers fabrics are "removed" all the time in ways that should not remove anything from persistent storage...